File: SAMPLES/LUA/test_scope.lua
-- test_scope.lua : test portee, pile et cadres (TODO 18)

-- Test 1 : do...end cree une portee locale
local a = 1
do
  local a = 2
  print(a)
end
print(a)

-- Test 2 : nettoyage de pile apres do...end
local b = 10
do
  local c = 20
  local d = 30
  print(c)
  print(d)
end
print(b)

-- Test 3 : do...end imbrique
do
  local x = 5
  do
    local y = 6
    print(x)
    print(y)
  end
  print(x)
end

-- Test 4 : fonction avec sauvegarde SI/DI
function compute(n)
  local r = n + 1
  return r
end
print(compute(10))

-- Test 5 : variable locale capturee (upvalue)
local val = 100
function show_val()
  print(val)
end
show_val()

-- Test 6 : portee dans boucle for
local sum = 0
for i = 1, 3 do
  local tmp = i
  sum = sum + tmp
end
print(sum)

-- Test 7 : do...end vide
do
end
print(42)

File: SAMPLES/KOTLIN/test_scope.kt
// test_scope.kt - Test portee, pile et cadres (TODO 18)

// Test 1 : variables locales dans une fonction
fun add(a: Int, b: Int): Int {
    val sum = a + b
    return sum
}

println(add(3, 7))

// Test 2 : portee de bloc - variable locale dans un if
val x = 10
if (x > 5) {
    val localVar = x + 1
    println(localVar)
}
println(x)

// Test 3 : portee de bloc dans une boucle while
var counter = 0
while (counter < 3) {
    val temp = counter * 2
    println(temp)
    counter = counter + 1
}

// Test 4 : portee de bloc dans une boucle for
for (i in 1..3) {
    val doubled = i * 2
    println(doubled)
}

// Test 5 : fonction avec plusieurs variables locales
fun compute(a: Int, b: Int): Int {
    val sum = a + b
    val diff = a - b
    val product = sum * diff
    return product
}

println(compute(5, 3))

// Test 6 : fonctions imbriquees - registres SI DI preserves
fun outer(n: Int): Int {
    val a = n + 1
    return a * 2
}

println(outer(4))

// Test 7 : portee else
val y = 3
if (y > 10) {
    println(99)
} else {
    val msg = y + 100
    println(msg)
}
